checkQC/handlers/unidentified_index_handler.py

```diff
@@ -211,34 +264,50 @@ def __eq__(self, other):
         def __hash__(self):
             return hash(self.__class__.__name__ + self.search_index + self.found_sample + self.found_lane)
 
+    def __init__(self, samplesheet):
+        self.samplesheet_dict = self.transform_samplesheet_to_dict(samplesheet)
+
     @staticmethod
-    def index_in_samplesheet(samplesheet_dict, index):
+    def transform_samplesheet_to_dict(samplesheet):
+        """
+        Transform samplesheet to dict on form {index -> {lane -> sample name}}
+        :param samplesheet:
+        :return: dict of index, lane and sample name
+        """
+
+        samplesheet_dict = defaultdict(lambda: defaultdict(dict))
+        for s in samplesheet:
+            index = s['index']
+            if s.get('index2'):
+                index += '+{}'.format(s['index2'])
+            # MiSeqs for example do not have lanes in their samplesheets, thus we will default to using lane 1
+            # if nothing has been specified in the samplesheet. /JD 20181101
+            lane = s.get('Lane', '1')
+            samplesheet_dict[index][lane] = s['Sample_Name']
+
+        return samplesheet_dict
+
+    def exact_index_in_samplesheet(self, index):
         """
         Search the sample sheet dict for an index to find which lane and which sample
         it is associated with.
-        :param samplesheet_dict:
         :param index:
-        :return:
+        :return: generator of SearchHits
         """
-        index_hits = samplesheet_dict.get(index)
+        index_hits = self.samplesheet_dict.get(index)
         if index_hits:
             for lane, sample in index_hits.items():
-                yield UnidentifiedIndexHandler.SearchHit(index, sample, lane)
+                yield _SamplesheetSearcher.SearchHit(index, sample, lane)
 
-    def number_of_reads_per_lane(self):
+    def one_index_match_in_samplesheet(self, index):
         """
-        Transform conversion results into dict of lane -> total clusters pass filer
-        :return: dict {<lane>: <total clusters pass filer>}
+        Search for in-exact matches in samplesheet, i.e. hits where there is a hit, for one, but not
+        both indexes in a dual index tag.
+        :param index:
+        :return:
         """
-        nbr_of_reads_per_lane = {}
-        for lane_dict in self.conversion_results:
-            nbr_of_reads_per_lane[int(lane_dict["LaneNumber"])] = int(lane_dict["TotalClustersPF"])
-        return nbr_of_reads_per_lane
 
-    @staticmethod
-    def get_complementary_sequence(seq):
-        conv_table = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G', 'N': 'N', '+': '+'}
-        comp_seq = ''
-        for c in seq:
-            comp_seq += conv_table[c]
-        return comp_seq
+        for samplesheet_index in self.samplesheet_dict.keys():
+            if index in samplesheet_index:
```

Comment:
perhaps you want to ensure that you are not matching a partial index here? i.e. something like
`if index in samplesheet_index.split('+'):`

Reply:
I've added a change to fix this, in the commit f741a80 have a look at what you think.

Reply:
looks good to me, do you want me to merge?